python/grass/tools/support.py

```diff
@@ -0,0 +1,276 @@
+                " (check session setup and documentation for tool and attribute names)"
+            )
+            raise exception_type(msg)
+        return tool_name
+
+    def get_function(self, name, exception_type):
+        tool_name = self.get_tool_name(name, exception_type)
+
+        def wrapper(**kwargs):
+            return self._run_function(tool_name, **kwargs)
+
+        wrapper.__doc__ = f"Run {tool_name} as function"
+
+        return wrapper
+
+    def __getattr__(self, name):
+        return self.get_function(name, exception_type=AttributeError)
+
+    @staticmethod
+    def levenshtein_distance(text1: str, text2: str) -> int:
+        if len(text1) < len(text2):
+            return ToolFunctionResolver.levenshtein_distance(text2, text1)
+
+        if len(text2) == 0:
+            return len(text1)
+
+        previous_row = list(range(len(text2) + 1))
+        for i, char1 in enumerate(text1):
+            current_row = [i + 1]
+            for j, char2 in enumerate(text2):
+                insertions = previous_row[j + 1] + 1
+                deletions = current_row[j] + 1
+                substitutions = previous_row[j] + (char1 != char2)
+                current_row.append(min(insertions, deletions, substitutions))
+            previous_row = current_row
+
+        return previous_row[-1]
+
+    def suggest_tools(self, text):
+        """Suggest matching tool names based on provided text.
+
+        The first five tools matching the condition are returned,
+        so this always returns at most five names, but it may not be the best matches.
+        The returned names are sorted alphabetically (not by priority).
+        This specific behavior may change in the future versions.
+        """
+        result = []
+        max_suggestions = 5
+        for name in self.names():
+            if ToolFunctionResolver.levenshtein_distance(text, name) < len(text) / 2:
+                result.append(name)
+            if len(result) >= max_suggestions:
+                break
+        result.sort()
+        return result
+
+    def names(self):
+        if self._names:
+            return self._names
+        self._names = [name.replace(".", "_") for name in gs.get_commands()[0]]
+        return self._names
+
+
+class ToolResult:
+    """Result returned after executing a tool"""
+
+    def __init__(self, *, name, command, kwargs, returncode, stdout, stderr):
+        self._name = name
+        self._command = command
+        self._kwargs = kwargs
+        self.returncode = returncode
+        self._stdout = stdout
+        self._stderr = stderr
+        self._text = None
+        self._cached_json = None
+
+    @property
+    def text(self) -> str | None:
+        """Text output as decoded string"""
+        if self._text is not None:
+            return self._text
+        if self._stdout is None:
+            return None
+        if isinstance(self._stdout, bytes):
+            decoded_stdout = gs.decode(self._stdout)
```

Comment:
The bytes case isn't tested (yet), maybe on Windows it will and I'm just too early.

Otherwise, maybe harder to test if we don't really use bytes anywhere.